File: Coding-Patterns/Coding-Patterns-Python/02_trees/level_order_traversal.py
# LeetCode 102 - Binary Tree Level Order Traversal
# Difficulty: Medium
# Pattern: BFS with queue
#
# Problem: Return the level-order traversal of a binary tree's node values
# (i.e., left to right, level by level).
#
# Approach: BFS using a deque. At each level, process exactly len(queue)
# nodes (snapshot of current level size before adding children).
#
# Time: O(n)  Space: O(n)

from typing import Optional, List
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def level_order(root: Optional[TreeNode]) -> List[List[int]]:
    if not root:
        return []
    result = []
    queue = deque([root])
    while queue:
        level_size = len(queue)
        level = []
        for _ in range(level_size):
            node = queue.popleft()
            level.append(node.val)
            if node.left:  queue.append(node.left)
            if node.right: queue.append(node.right)
        result.append(level)
    return result


# ---- Tests ----
if __name__ == "__main__":
    #     3
    #    / \
    #   9  20
    #     /  \
    #    15   7
    root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    assert level_order(root) == [[3], [9, 20], [15, 7]]
    assert level_order(None) == []
    print("All tests passed.")

File: Coding-Patterns/Coding-Patterns-CSharp/CodingPatterns/02_trees/LevelOrderTraversal.cs
// LeetCode 102 - Binary Tree Level Order Traversal
// Difficulty: Medium
// Pattern: BFS with queue
//
// Time: O(n)  Space: O(n)

namespace CodingPatterns.Trees;

public class LevelOrderTraversal
{
    public IList<IList<int>> LevelOrder(TreeNode root)
    {
        var result = new List<IList<int>>();
        if (root == null) return result;

        var queue = new Queue<TreeNode>();
        queue.Enqueue(root);

        while (queue.Count > 0)
        {
            int levelSize = queue.Count;
            var level = new List<int>();
            for (int i = 0; i < levelSize; i++)
            {
                var node = queue.Dequeue();
                level.Add(node.Val);
                if (node.Left != null)  queue.Enqueue(node.Left);
                if (node.Right != null) queue.Enqueue(node.Right);
            }
            result.Add(level);
        }
        return result;
    }

    // ---- Tests ----
    public static void Run()
    {
        var sol = new LevelOrderTraversal();

        //      3
        //     / \
        //    9  20
        //      /  \
        //     15   7
        var root = new TreeNode(3,
            new TreeNode(9),
            new TreeNode(20, new TreeNode(15), new TreeNode(7)));

        foreach (var level in sol.LevelOrder(root))
            Console.WriteLine($"[{string.Join(",", level)}]");   // [3] [9,20] [15,7]

        Console.WriteLine(sol.LevelOrder(null).Count);           // 0
    }
}
